File: sdks/typescript/src/resources/model-routes.ts
import type { ModelConnection, ModelRoute, ModelRouteRevision } from "../generated/types.ts";
import type { Palai } from "../client.ts";
import { callArgs, enc, type CallOptions, type ListView } from "./shared.ts";

// The projections are the generated contract types (ModelConnection/ModelRoute/ModelRouteRevision) — no
// more hand-written open interfaces (E16 T1 retires the E13 T10 `[key: string]: unknown` shape). The
// request params below shape a write body, so they stay local.

// A connection binds a provider family to a secret REF, never a value — a request that inlines a
// credential is rejected by the server's strict decode.
export interface ModelConnectionCreateParams {
  provider: string;
  secret_ref: string;
  /**
   * This connection's OWN endpoint (E29). REQUIRED by the `openai-compatible` family — that family has no
   * endpoint of its own, and an empty one would dial the OpenAI API with a key minted for something else —
   * and refused on families that do have one. Omit for `provider-one` / `provider-two`.
   */
  base_url?: string;
}

/**
 * One credential probe's finding (E29). `outcome` is the verdict; `not_probed` means NOTHING was checked
 * and must never be rendered as a pass. `checked` is the server's own sentence about what a probe does not
 * establish — the model id's validity above all — carried in the payload so a UI cannot overstate it.
 */
export interface ModelConnectionVerification {
  object: string;
  connection_id: string;
  provider: string;
  outcome: "credential_accepted" | "credential_rejected" | "unreachable" | "transient" | "not_probed";
  status?: number;
  endpoint?: string;
  detail?: string;
  checked?: string;
}
export interface ModelRouteCreateParams {
  name: string;
}
export interface ModelRouteRevisionCreateParams {
  model: string;
  connection_id: string;
}

// ModelRoutes is the model-routing admin surface (E13 T8 write + E16 T1 read-back, MCI-006): a project
// binds its own provider connection and publishes route revisions, so two projects on one stack run
// different models on different credentials — AND reads them back. The list methods return the admin
// ListView envelope ({object:"list", data:[…]}) — a full, small, tenant-scoped set, no cursor. Requires a
// key with the `provision` capability. A connection projection carries the secret REF name only, never a
// value.
export class ModelRoutes {
  #client: Palai;
  constructor(client: Palai) {
    this.#client = client;
  }

  // createConnection binds a provider family to a secret-ref handle (201).
  async createConnection(params: ModelConnectionCreateParams, options: CallOptions = {}): Promise<ModelConnection> {
    const r = await this.#client.request<ModelConnection>("POST", "/v1/model-connections", { body: params, ...callArgs(options) });
    return r.body;
  }

  // verifyConnection performs a REAL credential probe against the connection's endpoint (E29): the control
  // plane redeems this connection's own credential and asks the endpoint whether it accepts it. It resolves
  // (200) for every outcome it reached a verdict on, INCLUDING a rejected credential — the request
  // succeeded and the verdict is in the body, because a 4xx would blame the caller's request for the
  // operator's key. Only an unknown connection rejects, with a 404.
  async verifyConnection(connectionID: string, options: CallOptions = {}): Promise<ModelConnectionVerification> {
    const r = await this.#client.request<ModelConnectionVerification>("POST", `/v1/model-connections/${enc(connectionID)}/verify`, callArgs(options));
    return r.body;
  }

  // createRoute opens a named route alias for the project (201).
  async createRoute(params: ModelRouteCreateParams, options: CallOptions = {}): Promise<ModelRoute> {
    const r = await this.#client.request<ModelRoute>("POST", "/v1/model-routes", { body: params, ...callArgs(options) });
    return r.body;
  }

  // createRevision adds a DRAFT revision to a route (201) — a draft never steers a run; publishing does.
  async createRevision(routeID: string, params: ModelRouteRevisionCreateParams, options: CallOptions = {}): Promise<ModelRouteRevision> {
    const r = await this.#client.request<ModelRouteRevision>("POST", `/v1/model-routes/${enc(routeID)}/revisions`, { body: params, ...callArgs(options) });
    return r.body;
  }

  // publishRevision makes a draft revision the project's routed target. Re-publishing is idempotent.
  async publishRevision(routeID: string, revisionID: string, options: CallOptions = {}): Promise<ModelRouteRevision> {
    const r = await this.#client.request<ModelRouteRevision>("POST", `/v1/model-routes/${enc(routeID)}/revisions/${enc(revisionID)}/publish`, callArgs(options));
    return r.body;
  }

  // --- E16 T1 read-back (the E13 T10 write-only gap) ------------------------------------------------

  // listConnections returns the project's connections (secret REF names only, never values).
  async listConnections(options: CallOptions = {}): Promise<ListView<ModelConnection>> {
    const r = await this.#client.request<ListView<ModelConnection>>("GET", "/v1/model-connections", callArgs(options));
    return r.body;
  }

  // getConnection reads one connection by id; an absent/foreign id is a 404.
  async getConnection(connectionID: string, options: CallOptions = {}): Promise<ModelConnection> {
    const r = await this.#client.request<ModelConnection>("GET", `/v1/model-connections/${enc(connectionID)}`, callArgs(options));
    return r.body;
  }

  // listRoutes returns the project's route aliases.
  async listRoutes(options: CallOptions = {}): Promise<ListView<ModelRoute>> {
    const r = await this.#client.request<ListView<ModelRoute>>("GET", "/v1/model-routes", callArgs(options));
    return r.body;
  }

  // getRoute reads one route alias by id; an absent/foreign id is a 404.
  async getRoute(routeID: string, options: CallOptions = {}): Promise<ModelRoute> {
    const r = await this.#client.request<ModelRoute>("GET", `/v1/model-routes/${enc(routeID)}`, callArgs(options));
    return r.body;
  }

  // listRevisions returns a route's revisions (each with its derived `published` flag).
  async listRevisions(routeID: string, options: CallOptions = {}): Promise<ListView<ModelRouteRevision>> {
    const r = await this.#client.request<ListView<ModelRouteRevision>>("GET", `/v1/model-routes/${enc(routeID)}/revisions`, callArgs(options));
    return r.body;
  }

  // getRevision reads one revision of a route; an absent/foreign route or revision is a 404.
  async getRevision(routeID: string, revisionID: string, options: CallOptions = {}): Promise<ModelRouteRevision> {
    const r = await this.#client.request<ModelRouteRevision>("GET", `/v1/model-routes/${enc(routeID)}/revisions/${enc(revisionID)}`, callArgs(options));
    return r.body;
  }
}

File: sdks/go/modelroutes.go
package palai

import (
	"context"
	"net/http"
)

// ModelRoutes is the model-routing admin surface (E13 T8 write + E16 T1 read-back, MCI-006): a
// project binds its own provider connection and publishes route revisions, then reads them back.
// The list methods return the admin ListView envelope ({object:"list", data:[…]}) — a full, small,
// tenant-scoped set, no cursor. Requires a key with the `provision` capability. A connection
// projection carries the secret REF name only, never a value.
type ModelRoutes struct{ client *Client }

// ModelConnectionCreateParams binds a provider family to a secret REF (never a value — a request
// that inlines a credential is rejected by the server's strict decode).
type ModelConnectionCreateParams struct {
	Provider  string `json:"provider"`
	SecretRef string `json:"secret_ref"`
}

// ModelRouteCreateParams opens a named route alias.
type ModelRouteCreateParams struct {
	Name string `json:"name"`
}

// ModelRouteRevisionCreateParams adds a draft revision to a route.
type ModelRouteRevisionCreateParams struct {
	Model        string `json:"model"`
	ConnectionID string `json:"connection_id"`
}

// --- write surface (E13 T8) ------------------------------------------------------------------

// CreateConnection binds a provider family to a secret-ref handle (201).
func (m *ModelRoutes) CreateConnection(ctx context.Context, params ModelConnectionCreateParams, opts ...CallOption) (*ModelConnection, error) {
	var out ModelConnection
	return ptrOrErr(&out, m.post(ctx, "/v1/model-connections", params, &out, opts))
}

// CreateRoute opens a named route alias for the project (201).
func (m *ModelRoutes) CreateRoute(ctx context.Context, params ModelRouteCreateParams, opts ...CallOption) (*ModelRoute, error) {
	var out ModelRoute
	return ptrOrErr(&out, m.post(ctx, "/v1/model-routes", params, &out, opts))
}

// CreateRevision adds a DRAFT revision to a route (201) — a draft never steers a run; publishing does.
func (m *ModelRoutes) CreateRevision(ctx context.Context, routeID string, params ModelRouteRevisionCreateParams, opts ...CallOption) (*ModelRouteRevision, error) {
	var out ModelRouteRevision
	return ptrOrErr(&out, m.post(ctx, "/v1/model-routes/"+escapePathSegment(routeID)+"/revisions", params, &out, opts))
}

// PublishRevision makes a draft revision the project's routed target. Re-publishing is idempotent.
func (m *ModelRoutes) PublishRevision(ctx context.Context, routeID, revisionID string, opts ...CallOption) (*ModelRouteRevision, error) {
	var out ModelRouteRevision
	path := "/v1/model-routes/" + escapePathSegment(routeID) + "/revisions/" + escapePathSegment(revisionID) + "/publish"
	o := requestOptions{idempotent: true}
	applyCallOptions(&o, opts)
	return ptrOrErr(&out, m.client.doJSON(ctx, http.MethodPost, path, o, &out))
}

// --- read-back (E16 T1: the E13 T10 write-only gap) ------------------------------------------

// ListConnections returns the project's connections (secret REF names only, never values).
func (m *ModelRoutes) ListConnections(ctx context.Context, opts ...CallOption) (*ListView[ModelConnection], error) {
	var out ListView[ModelConnection]
	return ptrOrErr(&out, m.get(ctx, "/v1/model-connections", &out, opts))
}

// GetConnection reads one connection by id; an absent/foreign id is a 404.
func (m *ModelRoutes) GetConnection(ctx context.Context, connectionID string, opts ...CallOption) (*ModelConnection, error) {
	var out ModelConnection
	return ptrOrErr(&out, m.get(ctx, "/v1/model-connections/"+escapePathSegment(connectionID), &out, opts))
}

// ListRoutes returns the project's route aliases.
func (m *ModelRoutes) ListRoutes(ctx context.Context, opts ...CallOption) (*ListView[ModelRoute], error) {
	var out ListView[ModelRoute]
	return ptrOrErr(&out, m.get(ctx, "/v1/model-routes", &out, opts))
}

// GetRoute reads one route alias by id; an absent/foreign id is a 404.
func (m *ModelRoutes) GetRoute(ctx context.Context, routeID string, opts ...CallOption) (*ModelRoute, error) {
	var out ModelRoute
	return ptrOrErr(&out, m.get(ctx, "/v1/model-routes/"+escapePathSegment(routeID), &out, opts))
}

// ListRevisions returns a route's revisions (each with its derived `published` flag).
func (m *ModelRoutes) ListRevisions(ctx context.Context, routeID string, opts ...CallOption) (*ListView[ModelRouteRevision], error) {
	var out ListView[ModelRouteRevision]
	return ptrOrErr(&out, m.get(ctx, "/v1/model-routes/"+escapePathSegment(routeID)+"/revisions", &out, opts))
}

// GetRevision reads one revision of a route; an absent/foreign route or revision is a 404.
func (m *ModelRoutes) GetRevision(ctx context.Context, routeID, revisionID string, opts ...CallOption) (*ModelRouteRevision, error) {
	var out ModelRouteRevision
	path := "/v1/model-routes/" + escapePathSegment(routeID) + "/revisions/" + escapePathSegment(revisionID)
	return ptrOrErr(&out, m.get(ctx, path, &out, opts))
}

// --- small shared plumbing -------------------------------------------------------------------

func (m *ModelRoutes) post(ctx context.Context, path string, body, out any, opts []CallOption) error {
	o := requestOptions{body: body}
	applyCallOptions(&o, opts)
	return m.client.doJSON(ctx, http.MethodPost, path, o, out)
}

func (m *ModelRoutes) get(ctx context.Context, path string, out any, opts []CallOption) error {
	o := requestOptions{}
	applyCallOptions(&o, opts)
	return m.client.doJSON(ctx, http.MethodGet, path, o, out)
}

func applyCallOptions(o *requestOptions, opts []CallOption) {
	for _, opt := range opts {
		opt(o)
	}
}

// ptrOrErr returns (nil, err) on error, else (out, nil) — a one-liner for the resource getters.
func ptrOrErr[T any](out *T, err error) (*T, error) {
	if err != nil {
		return nil, err
	}
	return out, nil
}
